### PR TITLE
Nondeterministic tests #952

### LINKED ISSUE #952: Nondeterministic tests
Any time you get a nondeterministic test, put the relevant output here. We'll slowly try to eliminate all of them.

### PR DESCRIPTION
**Pull request by dariusf**
Fixes #952.

On hindsight, accumulating test output in an issue didn't really do anything to fix the problems.

This fixes a few unstable tests and lays the groundwork for fixing the remaining ones. I'll make issues for each of the latter group. In future, we should probably make issues for individual unstable tests and stable ones that we see failing randomly.

Most of them weren't being synchronised properly. The remaining unstable ones have deeper issues than just lack of synchronisation.

### CODE REVIEW

File: src/test/java/unstable/BoardTests.java [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+        pushKeys(KeyCode.ESCAPE);
+        waitUntilNodeDisappears("#boardnameinput");
+    }
+
+//    private void selectNthMenuItem(int n) {
+//        for (int i = 0; i < n; i++) {
+//            push(KeyCode.DOWN);
+//        }
+//        push(KeyCode.ENTER);
+//    }
```
Comment (damithc) [inline src/test/java/unstable/BoardTests.java]: remove?

File: src/test/java/unstable/BoardTests.java [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+    public void savingBoard() {
+
+        PanelControl panelControl = TestController.getUI().getPanelControl();
+
+        saveBoardWithName("Board 1");
+        pushKeys(CREATE_RIGHT_PANEL);
+        awaitCondition(() -> 2 == panelControl.getPanelCount());
+
+        traverseMenu("Boards", "Open", "Board 1");
+//        selectNthMenuItem(1);
```
Comment (damithc) [inline src/test/java/unstable/BoardTests.java]: more commented lines.

Comment (dariusf) [inline src/test/java/unstable/BoardTests.java]: Yes, I'm not done with this yet. Trying to get everything to pass first.

**Comment by dariusf** [comment]
@ndt could you help me to review this?

File: src/test/java/guitests/UITest.java [hunk 3]
```diff
@@ -362,6 +358,14 @@ private void awaitCondition(Callable<Boolean> condition, int timeoutInSeconds) {
     }
 
     /**
+     * Captures the common pattern of waiting for the result of a function,
+     * then asserting that it is equal to some value.
+     */
+    public <T> void waitAndAssert(T expected, Supplier<T> actual) {
+        awaitCondition(() -> expected.equals(actual.get()));
+    }
```
Comment (ndt93) [inline src/test/java/guitests/UITest.java]: This method should better be called `waitAndAssertEquals`?

File: src/test/java/guitests/UITest.java [hunk 4]
```diff
@@ -375,14 +379,58 @@ public void clickMenu(String... menuNames) {
...
+                    current = ((Menu) current).getItems().stream()
+                        .filter(m -> m.getText().equals(names[j]))
+                        .findFirst()
+                        .orElseThrow(() ->
+                            new IllegalArgumentException(
+                                String.format("%s is not a valid menu item", names[j])));
+                } else {
+                    throw new IllegalArgumentException(
+                        String.format("Menu %s is not as nested as arguments require", names[0]));
+                }
```
Comment (ndt93) [inline src/test/java/guitests/UITest.java]: We can use `if (!(current instanceof Menu)`)` and move the `throw` statement up so we can eliminate the `else` clause.

**Comment by dariusf** [comment]
Made the changes.

**Comment by ndt93** [comment]
Looks good for me. @damithc any comments?

File: src/test/java/guitests/UITest.java [hunk 5]
```diff
@@ -362,6 +358,14 @@ private void awaitCondition(Callable<Boolean> condition, int timeoutInSeconds) {
     }
 
     /**
+     * Captures the common pattern of waiting for the result of a function,
```
Comment (damithc) [inline src/test/java/guitests/UITest.java]: ~~Captures the common pattern of waiting~~ Waits ?

File: src/test/java/unstable/BoardTests.java [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+        ui.updateTitle();
+    }
+
+    @Before
+    public void before() {
+        PlatformEx.runAndWait(BoardTests::reset);
+    }
+
+    @Test
+    public void creatingAndClosingPanels() {
```
Comment (damithc) [inline src/test/java/unstable/BoardTests.java]: Can follow the new naming convention for tests?

File: src/test/java/unstable/BoardTests.java [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+        traverseMenu("Boards", "Save");
+        waitUntilNodeAppears("#boardnameinput");
+        ((TextField) find("#boardnameinput")).setText("Board 1");
+        click("OK");
+
+        assertEquals(1, panelControl.getNumberOfSavedBoards());
+        assertEquals(ui.getTitle(), uiTitleWithBoard("Board 1"));
+    }
+
+    private static String uiTitleWithBoard(String boardName) {
```
Comment (damithc) [inline src/test/java/unstable/BoardTests.java]: Method name should start with `get`?

File: src/test/java/unstable/MetadataUpdateTest.java [hunk 8]
```diff
@@ -22,19 +18,21 @@ public void testUpdatedTriggersMetadata() {
         ensureMetadataDownloadIsTriggered();
         ensureMetadataIsReceived();
 
-        TestUtils.awaitCondition(() ->
-            findQuiet("1 comments since, involving test.").isPresent());
-        assertTrue(findQuiet("2 comments since, involving User 1, User 2.").isPresent());
+        awaitCondition(() -> existsQuiet("1 comments since, involving test."));
+        awaitCondition(() -> existsQuiet("2 comments since, involving User 1, User 2."));
+        // assertTrue(findQuiet().isPresent());
```
Comment (damithc) [inline src/test/java/unstable/MetadataUpdateTest.java]: remove commented code here and below?

**Comment by damithc** [comment]
Some cosmetic changes requested. 
@ndt93 next time, label as `toMerge` to send to my review.

**Comment by ndt93** [comment]
Ok noted.

**Comment by dariusf** [comment]
Okay, done. I didn't change the method names for some tests affected by this, because they require more extensive changes to structure and that should probably be done separately.

**Comment by ndt93** [comment]
@damithc Merge waiting to be approved.

**Pull request merged**